### ISSUE TITLE
RFC: Dynamic guides.xml adaptation for GitHub rendering action

### ISSUE THREAD

**Issue by garvinhicking** [TYPO3-Documentation/render-guides#158]
# Current situation

## gh-render-action

We have a GitHub action at https://github.com/TYPO3-Documentation/gh-render-action/blob/main/action.yml which provides a GitHub action step to render documentation.

## t3docs-ci-deploy

That action is referred to in the Repository https://github.com/TYPO3-Documentation/t3docs-ci-deploy/blob/main/.github/workflows/main-rendering.yml:

```
      - name: Render Repository
        uses: TYPO3-Documentation/gh-render-action@main
        id: rendering
        with:
          repository_url: ${{ github.event.client_payload.repository_url }}
          source_branch: ${{ github.event.client_payload.source_branch }}
          target_branch_directory: ${{ github.event.client_payload.target_branch_directory }}
```

The action takes 3 arguments (repository url, source branch, target branch directory for the deployment).

Whenever the rendering of a repository like https://github.com/TYPO3-Documentation/TYPO3CMS-Reference-CoreApi/tree/documentation-draft/Documentation is requested, this action gets executed with the payload input that specifies:

* The URL of the GIT repository to render (i.e. `https://github.com/TYPO3-Documentation/TYPO3CMS-Reference-CoreApi.git`)
* The GIT branch (i.e. `main` or `documentation-draft`) of the target repository to render
* The destination directory of the rendered documentation (i.e. `draft` - the part of `https://docs.typo3.org/m/typo3/reference-coreapi/***draft***/en-us`)

## back to gh-render-action

Those three parameters are specified in the trigger, get passed on to `gh-render-action` and prepares the local directory:

```
    - uses: TYPO3-Documentation/render-guides@main
      id: render-guides
      if: steps.enable_guides.outputs.GUIDES == 'true'
      with:
        working-directory: t3docsproject
        config: ./t3docsproject/Documentation/
        output: ./RenderedDocumentation/Result/project/0.0.0
```

The output directory is just temporary, the rest of the action actually takes care it gets put into the specified `${{ env.target_branch_directory }}`  directory tree.

**Please note**, at this point neither the variable `${{ env.source_branch }}` nor `${{ env.target_branch_directory }}` is passed (and evaluated) in the action `TYPO3-Documentation/render-guides@main`.

## render-guides GitHub action

This action step uses our Docker image, and uses `https://github.com/TYPO3-Documentation/render-guides/blob/main/action.yml` to actually run the rendering process. It uses this portion:

```
runs:
  using: 'docker'
  image: 'docker://ghcr.io/typo3-documentation/render-guides:main'
  env:
    SHELL_VERBOSITY: '3'
  args:
    - '--working-dir=${{ inputs.working-directory }}'
    - '--config=${{ inputs.config }}'
    - '--output=${{ inputs.output }}'
    - '${{ inputs.input }}'
```

This carries over the inputs from the gh-render-action:

* `working-directory` (`t3docsproject`)
* `config` (`./t3docsproject/Documentation/`)
* `output` (`./RenderedDocumentation/Result/project/0.0.0`)

It operates on the locally checked out repository inside the GitHub workflow, that uses the specified branch earlier (which is checked out into `/github/workspace/t3docsproject/`).

That means, the `/github/workspace/t3docsproject/Documentation/guides.xml` of the project will be evaluated and decide for the renderer HTML output which TOC tree and version toggle will be contained (from `https://docs.typo3.org/m/typo3/reference-coreapi/draft/en-us/`):

```
<div id="toc-version" class="toc-version">
  <span class="toc-version-prefix">Release:</span>
  <span class="toc-version-number">***draft***</span>
</div>
```

The reason that "draft" shows up at this place is because https://github.com/TYPO3-Documentation/TYPO3CMS-Reference-CoreApi/blob/documentation-draft/Documentation/guides.xml specifies:

```
<extension ... edit-on-github-branch="documentation-draft" />
  <project ... version="draft" release="draft" />
```

## back to gh-render-action

Now the OLD workflows for sphinx-based rendering does this:

```
    - id: jobfile
      if: steps.enable_guides.outputs.GUIDES == 'false'
      working-directory: t3docsproject
      shell: bash
      run: |
        touch jobfile.json
        cat << EOF > jobfile.json
        {
            "Overrides_cfg": {
                "general": {
                    "release": "${{ env.target_branch_directory }}"
                },
                "html_theme_options": {
                    "docstypo3org": "yes"
                }
            }
        }
        EOF
```

This writes a temporary `jobfile.json`  in the working directory (not the repository) which is evaluated by the rendering, to override parts of the old `Settings.cfg`, so that a repository does not need to change `Settings.cfg` for each branch version, and maintain different variants of that file per branch. That makes branch creation easier.

# The Problem with this

As stated in the last paragraph above, the old rendering was able to enrich the rendering process with temporary state date only relevant to the GitHub action process to influence rendering.

Currently we only have these means:

* Static `Documentation/guides.xml` values of the project
* Static `/opt/guides/vendor/guides.xml` values of the `render-guides` repository

Plus, we're able to use the `packages/typo3-guides-cli/bin/typo3-guides configure ...` command to dynamically manipulate the local `guides.xml` file.

What we are lacking, is the ability to override per-repository `guides.xml` settings with GitHub action inputs.

Currently, we probably only need to overrite the following attributes of the `guides.xml` file:

* `<extension edit-on-github-branch>` (-> `${{ env.source_branch }}`)
* `<project version>` (-> `${{ env.target_branch_directory }}`)
* `<project release>` (-> `${{ env.target_branch_directory }})

(Note that both version+release by intent use the same variable content, that's also how it is with Sphinx currently)

# Ideas to solve this

I have a couple of ideas how we could manage this, but they need discussion and decission.

## Variant A: Provide and evaluate GitHub Environment variables

GitHub passes all existing custom GitHub environment variables on to the docker container process. So inside the process, the `render-guides` (and thus `guides`) project could actually read out the environment variables directly.

If those variables are found, we could override the specific values that are set in guides.xml in an additional step. I have not yet researched the actual place where this needs to be adapted, but I guess it needs to be done in the `https://github.com/phpDocumentor/guides/` repository. 

* Advantages
  * Probably easy to implement
  * In the future, more environment variables could be evaluated.
* Disadvantages
  * Reading/evaluating environment variables is a kind-of intransparent process to the user
  * it would influence other projects that could use this in the future
  * It makes the project configuration process even more complex due to these overrides

## Variant B: Provide and evaluate `/opt/guides/vendor/bin/guides` parameters

Instead of using environment variables, we could expose this more specifically to the CLI of the `guides` as parameters.

Currently the CLI has these parameters:

* `output`
* `input-file`
* `input-format`
* `output-format`
* `log-path`
* `working-dir`
* `config`
* `theme`
* `fail-on-log`
* `progress`

So additionally, we could add:

* `config-branch`
* `config-project-release`
* `config-project-version`

* Advantages
  * Probably easy to implement
  * Command line arguments are clearly exposed and documented
* Disadvantages
  * Requires adaptation of GitHub action.yml files to pass through all the arguments
  * it would influence other projects that could use this in the future
  * It makes the project configuration process even more complex due to these overrides
  * It adds 3 more command line options and in the end people might wonder if they need to specify command line arguements or adapt the `guides.xml` file.

## Variant C: Enforce full per-repository `guides.xml` settings

We could also NOT go the route of supporting dynamic attributes. Then we would force people into writing custom directives per GIT-branch into their repositories.

* Advantages
  * No "hacks" needed for the current rendering approach
  * Clear instructions, single-source-of-truth is guides.xml
* Disadvantages
  * Drifting GIT branches, people may forget to adapt `guides.xml` for the rendering
  * GitHub actions would specify release versions that would NOT be evaluated. People might wonder why their GitHub action settings do not take effect, and may not recognize that `guides.xml` is the source-of-truth.

## Variant D: Dynamically alter `guides.xml` within the GitHub workflow

We would be able to execute dynamic `guides.xml` alteration as a new step within the workflow. Something like this in the `action.yml` file of `gh-render-action`, just before the current step `id: render-guides`:

````
    - uses: TYPO3-Documentation/render-guides@main
      id: adapt-render-guides
      if: steps.enable_guides.outputs.GUIDES == 'true'
      with:
        config-branch: '${{ env.source_branch }}'
        config-project-release: '${{ env.target_branch_directory }}'
        config-project-version: '${{ env.target_branch_directory }}'
````

and then in the `action.yml` file of `render-guides` we would need something along the lines of this:

````
runs:
  using: 'docker'
  image: 'docker://ghcr.io/typo3-documentation/render-guides:main'
  env:
    SHELL_VERBOSITY: '3'
  shell: bash
  working-directory: ${{ inputs.working-directory }}
  steps:
    - name: Adapt guides.xml with GitHub inputs
      run: |
        configure --project-version='${{ inputs.config-project-version }}' \
          --project-release='${{ inputs.config-project-release }}' \
          --extension-edit-on-github='${{ inputs.config-branch }}' \
    - name: Execute rendering within working directory and configured output.
      run: |
        render --working-dir=${{ inputs.working-directory }} \
          --config=${{ inputs.config }} \
          --output=${{ inputs.output }} \
          ${{ inputs.input }}
````

The above is untested and would need further adaptation of the `configure` CLI command, too.

* Advantages
  * Doesn't touch the internal logic of guides.xml is the single-source-of-truth
  * Integrates nicely as a step in the GitHub workflow
  * Can be used for much more adaption in the future, can also be used by end-users to change their repository's guides.xml, would allow this `configure` tool to become more fleshed out
* Disadvantages
  * Adds complexity to the GitHub workflow (and thus breaking potential)
  * I'm still getting myself into adapting GitHub workflows and I may make mistakes, or the idea itself isn't good and breaks concepts
  * The "configure" CLI command is yet more a "proof of concept" and would need thorough testing, and probably refactoring.
  * This does "magic" do the `guides.xml` of a repository, and it's harder to understand than specifying command line arguments or environment arguments as override input

## Variant E: Add another "guides-override.xml" evaluation

Added to the evaluation of the vendor `guides.xml` plus `Documentation/guides.xml` a third place could be evaluated (like the `jobfile` before). We could create a file like `guides-override.xml` dynamically, and it could dynamically overload any configuration setting, just like the `Documentation/guides.xml` file could do.

Maybe it could also be a simpler JSON file instead, to ease the dynamic creation of the file in the GitHub workflow (otherwise, all the XML schema information needs to be redundantly added in the workflow file, or we would again need a dynamic file creation tool like the variant before, which kind of defeats this variant as an alternative).

That file would not be kept in repositories, but only be created in working directories.

* Advantages
  * Probably easy to implement (into the `render` project, because that is where the `guides.xml` evaluation logic is in)
  * Similar to the current process
* Disadvantages
  * Adds complexity with a THIRD file of overloading configuration.
  * Also needs GitHub workflow adaptation for the file creation (with XML format, which would be hardcoded into the GitHub workflow - with JSON format, another evaluation logic would need to be added)

## Other Variants

I may be missing other ideas, open to futher variant proposals.

# TL;DR & my favorite variant

We want to be able to let the GitHub-action-based rendering process override a restricted set of variables (release, version, edit-on-github) on the fly.

There are several ideas to implement this, each with advantages and disadvantages.

Personally, I would like to implement Variant D. It uses an existing concept (which is not fully fleshed out yet though), it allows for much flexibility for any other override, it doesn't need to touch the `render` base project, and would enhance a tool that users could find very helpful.

**Comment by jaapio** [comment]
## gh-render-action

I would like to get rid of this repository at some point, once 
all projects are migrated to the new render-guides repo. 
It doesn't make sense to have a separate repo, since the only 
step that is really needed is the `render-guides` action. 

We can move some plumbing to the `t3docs-ci-deploy` repo,
that will make it much easier for everyone to understand
what is needed. 

## Configure command

My idea when I started the implementation of the `configure command`
was to have a single command that can be used to dynamically 
update the guides.xml during the rendering process.
As part of the final workflow in `t3docs-ci-deploy`, just like 
the jobfile is now created. But by calling a different action. 

## my favorite

That makes me agree with you that variant D is the best option.

We can modify the current `render-guides` action to a composite action:
https://github.com/orgs/community/discussions/36861 which is somewhat 
the same as the `gh-render-action` repo.

This will require use to create 2 new actions:
- `configure-guides` action (adjusted version of the action.yaml file calling configure command)
- `render-guides` action (existing action.yaml file)
- composite action that calls both actions

We can keep this all in the same repo, which allows us to test everything in the pipelines of render-guides.

Suggested folder structure:
```
  actions/
    configure-guides/
      action.yaml
    render-guides/
      action.yaml

  action.yaml
```

The action.yaml at the root of the repo will be the composite action. Calling the other 2 actions, 
passing through the required parameters. As the action will be a composite action, we can use the 
working-dir parameter of github actions. And we could move some of the plumbing to the composite action.
which is now in the `gh-render-action` repo.

**Comment by garvinhicking** [comment]
Implemented via #169
